File: sdks/csharp/Examples/3DScene.cs
/**
 * 3D Scene Demo
 * 
 * Demonstrates 3D rendering with PBR materials, lighting, and post-processing.
 * 
 * Run with: dotnet run --project Examples/3DScene.cs
 */

using System;
using Windjammer.SDK;

namespace Windjammer.Examples
{
    class Scene3D
    {
        static void Main(string[] args)
        {
            var app = new App();

            app.AddStartupSystem(() =>
            {
                // Camera
                new Camera3D(
                    position: new Vector3(0, 5, 10),
                    lookAt: new Vector3(0, 0, 0),
                    fov: 60.0f
                );

                // Lights
                new PointLight { Position = new Vector3(5, 5, 5), Color = new Color(1.0f, 0.8f, 0.6f, 1.0f), Intensity = 2000.0f };
                new PointLight { Position = new Vector3(-5, 5, 5), Color = new Color(0.6f, 0.8f, 1.0f, 1.0f), Intensity = 1500.0f };
                new PointLight { Position = new Vector3(0, 10, -5), Color = new Color(1, 1, 1, 1), Intensity = 1000.0f };

                // Meshes with PBR materials
                Mesh.Cube(1.0f).WithMaterial(new Material
                {
                    Albedo = new Color(0.8f, 0.2f, 0.2f, 1.0f),
                    Metallic = 0.8f,
                    Roughness = 0.2f,
                    Emissive = new Color(0.5f, 0.1f, 0.1f, 1.0f)
                });

                Mesh.Sphere(1.0f, 32).WithMaterial(new Material
                {
                    Albedo = new Color(0.2f, 0.2f, 0.8f, 1.0f),
                    Metallic = 0.5f,
                    Roughness = 0.5f,
                    Emissive = new Color(0.1f, 0.1f, 0.5f, 1.0f)
                });

                Mesh.Plane(10.0f).WithMaterial(new Material
                {
                    Albedo = new Color(0.3f, 0.3f, 0.3f, 1.0f),
                    Metallic = 0.0f,
                    Roughness = 0.9f
                });

                // Post-processing
                var post = new PostProcessing();
                post.EnableHDR(true);
                post.SetBloom(new BloomSettings { Threshold = 1.0f, Intensity = 0.8f, Radius = 4.0f, SoftKnee = 0.5f });
                post.SetSSAO(new SSAOSettings { Radius = 0.5f, Intensity = 1.5f, Bias = 0.025f, Samples = 16 });
                post.SetToneMapping(ToneMappingMode.ACES, 1.2f);
                post.SetColorGrading(new ColorGrading { Temperature = 0.1f, Tint = 0.0f, Saturation = 1.2f, Contrast = 1.1f });
            });

            app.AddSystem((Time time) =>
            {
                // Rotate objects for dynamic lighting
            });

            app.Run();
        }
    }
}

File: sdks/go/examples/3d_scene.go
// 3D Scene Demo
//
// Demonstrates 3D rendering with PBR materials, lighting, and post-processing.
//
// Run with: go run examples/3d_scene.go

package main

import (
	wj "github.com/windjammer/sdk-go/windjammer"
)

func main() {
	app := wj.NewApp()

	app.AddStartupSystem(func() {
		// Camera
		wj.NewCamera3D(
			wj.Vec3{X: 0, Y: 5, Z: 10},
			wj.Vec3{X: 0, Y: 0, Z: 0},
			60.0,
		)

		// Lights
		wj.NewPointLight(wj.Vec3{X: 5, Y: 5, Z: 5}, wj.Color{R: 1.0, G: 0.8, B: 0.6, A: 1.0}, 2000.0)
		wj.NewPointLight(wj.Vec3{X: -5, Y: 5, Z: 5}, wj.Color{R: 0.6, G: 0.8, B: 1.0, A: 1.0}, 1500.0)
		wj.NewPointLight(wj.Vec3{X: 0, Y: 10, Z: -5}, wj.Color{R: 1, G: 1, B: 1, A: 1}, 1000.0)

		// Meshes with PBR materials
		wj.MeshCube(1.0).WithMaterial(wj.Material{
			Albedo:   wj.Color{R: 0.8, G: 0.2, B: 0.2, A: 1.0},
			Metallic: 0.8,
			Roughness: 0.2,
			Emissive: wj.Color{R: 0.5, G: 0.1, B: 0.1, A: 1.0},
		})

		wj.MeshSphere(1.0, 32).WithMaterial(wj.Material{
			Albedo:   wj.Color{R: 0.2, G: 0.2, B: 0.8, A: 1.0},
			Metallic: 0.5,
			Roughness: 0.5,
			Emissive: wj.Color{R: 0.1, G: 0.1, B: 0.5, A: 1.0},
		})

		wj.MeshPlane(10.0).WithMaterial(wj.Material{
			Albedo:   wj.Color{R: 0.3, G: 0.3, B: 0.3, A: 1.0},
			Metallic: 0.0,
			Roughness: 0.9,
		})

		// Post-processing
		post := wj.NewPostProcessing()
		post.EnableHDR(true)
		post.SetBloom(wj.BloomSettings{Threshold: 1.0, Intensity: 0.8, Radius: 4.0, SoftKnee: 0.5})
		post.SetSSAO(wj.SSAOSettings{Radius: 0.5, Intensity: 1.5, Bias: 0.025, Samples: 16})
		post.SetToneMapping(wj.ToneMappingModeACES, 1.2)
		post.SetColorGrading(wj.ColorGrading{Temperature: 0.1, Tint: 0.0, Saturation: 1.2, Contrast: 1.1})
	})

	app.AddSystemWithTime(func(time *wj.Time) {
		// Rotate objects for dynamic lighting
	})

	app.Run()
}
